src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs

```diff
@@ -152,6 +163,58 @@ internal class CompletionHandler : IRequestHandler<CompletionParams, SumType<Com
             return result;
         }
 
+        private bool IsSimpleImplicitExpression(LSPDocumentSnapshot documentSnapshot, CompletionParams request, LanguageServerKind serverKind)
+        {
+            if (serverKind != LanguageServerKind.CSharp)
+            {
+                return false;
+            }
+
+            if (string.Equals(request.Context.TriggerCharacter, "@", StringComparison.Ordinal))
+            {
+                // Completion was triggered with `@` this is always a simple implicit expression
+                return true;
+            }
+
+            var snapshot = documentSnapshot.Snapshot;
+            var navigator = _textStructureNavigator.GetTextStructureNavigator(snapshot.TextBuffer);
```

Comment:
@AmadeusW would you mind reviewing how I'm utilizing the text structure navigator here to ensure I'm not doing anything awful 😄

Reply:
It looks good. You don't need to worry about caching or anything like that. 👍